patches/server/1058-Paper-Item-API.patch

```diff
@@ -0,0 +1,4571 @@
++        itemStack.handle.setTag(paperMeta.getPropertyHolder().getCompoundTag());
++        // Serialze the itemstack.
++        // We really don't care about the type/count, we just want to be able to serialize the main data in order to use the
++        // datafixer system
++        CompoundTag tag = new CompoundTag();
++        itemStack.handle.save(tag);
++        tag.remove("id"); // Remove material
++        tag.remove("Count"); // Remove count
++        meta.put("snbt", new SnbtPrinterTagVisitor().visit(tag));
++
++        return meta;
++    }
++
++    public static Map<String, Object> serialize(CraftMetaItem paperMeta) {
++        return serialize(convertBukkitToPaper(paperMeta));
++    }
++
++    public static PaperMeta convertBukkitToPaper(CraftMetaItem meta) {
++        CompoundTag compoundTag = new CompoundTag();
++        meta.apply(compoundTag);
++
++        return new PaperMeta(io.papermc.paper.property.nbt.NbtPropertyHolder.itemTagHolder(compoundTag));
++    }
++
++}
+diff --git a/src/main/java/io/papermc/paper/item/meta/LegacyMetaImpl.java b/src/main/java/io/papermc/paper/item/meta/LegacyMetaImpl.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..74e0ef051a8a27b3b8ef1f3a4cc48ee59681ef1d
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/item/meta/LegacyMetaImpl.java
+@@ -0,0 +1,158 @@
++package io.papermc.paper.item.meta;
++
++import com.destroystokyo.paper.Namespaced;
++import com.google.common.collect.Sets;
++import io.papermc.paper.property.nbt.NbtPropertyHolder;
++import net.kyori.adventure.text.Component;
++import net.kyori.adventure.text.format.TextDecoration;
++import net.kyori.adventure.text.serializer.gson.GsonComponentSerializer;
++import net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer;
++import net.md_5.bungee.api.chat.BaseComponent;
++import net.md_5.bungee.chat.ComponentSerializer;
++import org.bukkit.Material;
++import org.bukkit.craftbukkit.inventory.tags.DeprecatedCustomTagContainer;
++import org.bukkit.inventory.meta.ItemMeta;
++import org.bukkit.inventory.meta.tags.CustomItemTagContainer;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++import java.util.*;
++import java.util.stream.Collectors;
++
++import static io.papermc.paper.inventory.item.properties.ItemProperties.CAN_DESTROY_BLOCK_LIST;
++import static io.papermc.paper.inventory.item.properties.ItemProperties.CAN_PLACE_ON_BLOCK_LIST;
++
++// Done for readability
++interface LegacyMetaImpl extends ItemMeta {
++
++    @Override
++    default @NotNull String getDisplayName() {
++        return LegacyComponentSerializer.legacySection().serializeOr(this.displayName(), "");
++    }
++
++    @Override
++    default void setDisplayName(@Nullable String name) {
++        this.displayName(
++            LegacyComponentSerializer.legacySection().deserializeOr(name, Component.empty())
++                .decorationIfAbsent(TextDecoration.ITALIC, TextDecoration.State.FALSE)
```

Comment:
Is there a reason we are changing this to use adventure's serializer instead of just sticking with CraftChatMessage? Just feels unneeded and you have to handle specific quirks ourselves, like the italic stuff here.